### DISCUSSION TITLE
500 | Internal Server Error after successfully build and deploy

### DISCUSSION THREAD

**Discussion by mojo-KB** [vercel/next.js#43989] [Help] [answered]
Hello everyone,

So after successfully deploying my site, I got 500 Internal Server Errors. I don't know what went wrong. I was thinking maybe because of my environment variables which I have already stored in Vercel. Does anyone have any suggestions on which part I should look into?

**Comment by CTOverton** [comment]
Sorry to not add anything helpful, but I too have this issue when I deployed my Next.js site to AWS. The build works fine locally and in aws amplify it builds as well, but I'm not sure if the SSR is running. 

I would love to see if anyone has a build config or examples of hosted sites and how they did it.

**Reply by CTOverton** [reply]
Follow up to my issue, AWS while supporting Next.js 13 does not yet support the app directory beta. So I'll simply host with Vercel :)

**Comment by icyJoseph** [comment, marked as answer]
How are you getting these errors? Could you describe more about what your site does? 

500 errors are pretty much your server giving up or crashing, so without much context, we'll be here guessing for as long as the sun shines.

**Reply by mojo-KB** [reply]
This is how the site shows up. 

<img width="2869" alt="Screenshot 2022-12-13 at 12 15 58 PM (3)" src="https://user-images.githubusercontent.com/56409448/207400051-4c20f92d-a5f4-4f54-ba7e-7181ab959e16.png">

**Reply by icyJoseph** [reply]
Alright there's not enough info there. Are there any logs on the Vercel dashboard?

How's this page running, is it using GetStaticProps? GetServerSideProps?

**Reply by mojo-KB** [reply]
I used getServerSideProps to fetch data from Sanity CMS.
<img width="1412" alt="Screenshot 2022-12-13 at 1 00 41 PM (2)" src="https://user-images.githubusercontent.com/56409448/207410187-7f662b20-bd99-445a-bf81-858fe43918bb.png">

**Reply by icyJoseph** [reply]
Alright but then you should look for the function logs. Each deployment on Vercel has runtime logs per function. There you should see more information. It is live, so once you find it, you have to trigger a page view.

My guess is that the request towards Sanity is yielding a 5xx error itself and trying to JSON decode it trips the GSSP handler. The question is why would the Sanity API return error.

**Reply by icyJoseph** [reply]
Something like this:

![Screenshot 2022-12-13 at 22 07 40](https://user-images.githubusercontent.com/21013447/207443708-ef499eaa-b1fa-4233-9099-beb9862e2ea0.png)

**Reply by mojo-KB** [reply]
Okay, now I see there's an error. 
<img width="1381" alt="Screenshot 2022-12-13 at 10 28 35 PM (2)" src="https://user-images.githubusercontent.com/56409448/207498355-292f0748-5c42-4c8a-8c5c-cdabeab238d7.png">

**Reply by icyJoseph** [reply]
Alright that means that it is expecting a client to run on port 3000. 

Is that where you expect the sanity CMS endpoint to live at?

It is a bit odd that the experimental warning is labelled error, though, I don't think that's the issue.

I think the issue is the fetch failed error. It must be that the sanity CMS endpoint is not correctly configured.

Of course, even if the sanity cms is broken, we should try to render something.

You should wrap most of the async work of your GSSP function into a try catch, log the reason for failure, and return fallback props, so that you can show something to visitors. First thing first though, can you tell why is it trying to make a call :3000?

**Reply by mojo-KB** [reply]
I created internal api endpoints inside the api directory and created a utility function to fetch it back from localhost:3000/api.

_**./pages/index.tsx**_
```
export const getServerSideProps: GetServerSideProps<Props> = async () => {
  const data_pageInfo = await fetchPageInfo();
  const pageInfo: PageInfo = data_pageInfo.pageInfo;

  return {
    props: {
      pageInfo,
 
    }
  }
}
```



_**./utils/fetchPageInfo.ts**_
```
import { PageInfo } from "../typing";
export const fetchPageInfo = async () => {
    try {
       const res = await fetch(`${process.env.NEXT_PUBLIC_BASE_URL}/api/getPageInfo`, {
        method: "GET",
        headers: {
            accept: 'application/json'
        }
       });

       if (!res.ok) {
        throw new Error(`Error! Status: ${res.status}`)
       }
       const data = await res.json();
       return data;
    } catch (error) {
       console.log(error);
    }
}
```


_**./pages/api/getPageInfo.ts**_
```
import type { NextApiRequest, NextApiResponse } from "next";
import { groq } from "next-sanity"
import { sanityClient } from "../../sanity";
import { PageInfo } from "../../typing";

const query = groq`
    *[_type == "pageInfo"][0]
`

type Data = {
    pageInfo: PageInfo; 
}

export default async function handler(req: NextApiRequest, res: NextApiResponse<Data>) {
    const pageInfo: PageInfo= await sanityClient.fetch(query)

    res.status(200).json({ pageInfo })
}
```
_**inside my .env**_
NEXT_PUBLIC_BASE_URL=http://localhost:3000

**Reply by icyJoseph** [reply]
Alright, here we go...

So you used `fetch('/api')` from within `GSSP`, and it complained about using absolute URLs so you threw in the localhost bit to make it find itself locally.

The root issue is that reaching for `api` routers from within `GSSP` is at best, unnecessary. https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

What you are doing is flying a new request to a server, which is your own. What makes this worst is that you are assuming your app runs in localhost port 3000, but in truth Vercel will place it wherever, and in which ever port. In fact, GSSP might even run as a serverless lambda, and the result passed back to server. 

What you need to do is parametrise your sanity client usage, so that you can also call it from within `GSSP`.  You might be worried about leaking to client code, but this tool https://next-code-elimination.vercel.app/ shows what is actually packaged. 

In a sense, `GSSP` is already an specialised API route, the specialisation is that the results of `GSSP` are passed as props to the page component. 

> Also in general, servers should be, to the extend that's possible, agnostic to the domain where they run. 


```tsx
export const getServerSideProps: GetServerSideProps<Props> = async () => {
  const data_pageInfo = await sanityClient.fetch(query)
  const pageInfo: PageInfo = data_pageInfo.pageInfo;

  return {
    props: {
      pageInfo,
 
    }
  }
}
```

Or:

```tsx
import { groq } from "next-sanity"
import { type SanityClient } from "../../sanity"; // assuming this type exists 
// otherwise export type SanityClient = typeof sanityClient (within the sanity file)
import { PageInfo } from "../../typing";

const query = groq`
    *[_type == "pageInfo"][0]
`

export const fetchPageInfo = async (client: SanityClient) => {
    try {
       const pageInfo: PageInfo= await client.fetch(query)

      return pageInfo;
    } catch (error) {
       console.log(error);
    }
}
```

And then from GSSP:

```tsx
export const getServerSideProps: GetServerSideProps<Props> = async () => {
  const data_pageInfo = await fetchPageInfo(sanityClient);
  const pageInfo: PageInfo = data_pageInfo.pageInfo;

  return {
    props: {
      pageInfo,
 
    }
  }
}
```

And from `API` routes:

```tsx
import type { NextApiRequest, NextApiResponse } from "next";
// skipping imports

export default async function handler(req: NextApiRequest, res: NextApiResponse<Data>) {
    const pageInfo: PageInfo= await fetchPageInfo(sanityClient);

    res.status(200).json({ pageInfo })
}
```

Of course, assuming that you want to keep a public `api` route, from where other parties can consume the page info, otherwise, keep it all in GSSP.


## No No, I want to keep my things separated blah blah

I've seen this issue pop up a lot, and people always say that they want to keep their concerns separated and what not excuses to not organise the code properly through proper parametrisation. 

If that's your case then you can use VERCEL_URL, https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables:

> VERCEL_URL
> The URL of the deployment. Example: [my-site-7q03y4pi5.vercel.app](https://vercel.com/docs/concepts/projects/my-site-7q03y4pi5.vercel.app) 

So you'd do:

```tsx
export const fetchPageInfo = async () => {
    try {
       const res = await fetch(`${process.env.VERCEL_URL || "http://localhost:3000"}/api/getPageInfo`, {
        method: "GET",
        headers: {
            accept: 'application/json'
        }
       });

       if (!res.ok) {
        throw new Error(`Error! Status: ${res.status}`)
       }
       const data = await res.json();
       return data;
    } catch (error) {
       console.log(error);
    }
}
```

And don't forget to:

> To expose them to your Deployments, make sure Automatically expose System Environment Variables is checked in your Project Settings.

Again this is unnecessary, and far from optimal. Apologies if this section sounds a bit dry.

**Reply by mojo-KB** [reply]
It works!!! The problem was I did set the environment variable incorrectly. So I was supposed to set my environment variable NEXT_PUBLIC_BASE_URL which your example is VERCEL_URL to the domain, instead, I was setting it as http://localhost:3000. Thanks a lot.  🙌

**Reply by icyJoseph** [reply]
Great! Don't forget to mark as answered. Always happy to help.
